### ISSUE TITLE
Script Editor: In the "Method Filter" list, highlight the method the caret is currently inside of

### ISSUE THREAD

**Issue by Mickeon** [godotengine/godot-proposals#4073]
### Describe the project you are working on

Typical Godot game, using the Godot's built-in Script Editor for quite some time.

### Describe the problem or limitation you are having in your project

When switching across scripts,  **CTRL + Clicking** to go to method definitions, or even reloading the engine after a while, it takes a bit of time to figure out where in the script you currently are or last left off, particularly so when working on longer methods. The Mini-Map can only help so much as a nice visual clue, as it lacks the necessary information to figure that out.

Of course, larger scripts are always harder to navigate, but it does scale fairly poorly with "**Method Filter**" list. 

As brought up in a related discussion (https://github.com/godotengine/godot-proposals/discussions/4042), the "**Method Filter**" list could be more useful. It's often redundant and impractical to look at and make use of, despite it being prominently featured on the left side of the Script Editor, as part of the Scripts Panel. Due of its lack of detail, it's often more favourable to scroll across the script by other means as means of reference.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

**Highlight** the listed method the caret is currently inside of, _however way it is desirable_.
This will update whenever the caret is moved up or down the script.
Should the method name not be visible, the "**Method Filter**" list should synchronise and scroll accordingly until it is displayed.
No method name should be highlighted if not present in the current list.

E.g. the "**Method Filter**" list of this script will highlight "**push**" because the caret is inside `push()`
```
func push():
	sprite.play("pushed")
	if (so|und):
		sound.pitch_scale = rand_range(0.975, 1.025)
		sound.play()
```

This is one small way to turn the list into a more useful point of reference for users struggling to move around their code and getting lost.

Alternatively, instead of the caret's position, the scrolling position could be somehow used, but I wager figuring out where the user is actually looking at would be daunting, as well as yield plenty of inconsistent results. Should it be implemented this way, feel free to close the proposal.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Simple screenshot using my own code. Note that, as of currently, selecting a method on the list leaves it highlighted as such, until the script is changed. While this is most certainly be an oddity, using this exact highlight for the proposal could be ideal, although I do believe it stands out a bit too much while moving across the code.
![image](https://user-images.githubusercontent.com/66727710/155026007-3b274f16-cf76-4355-8662-b6a3967a2eee.png)

Alternative poor mock-up attempting to mimic the accent color as a highlight. Should be the same color used for the Script List's "temperature".
![image](https://user-images.githubusercontent.com/66727710/155025864-ed4fba37-e2a9-4e09-b923-dcd21151c207.png)

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not really.

### Is there a reason why this should be core and not an add-on in the asset library?

The Script Editor is very much core.

**Comment by lostminds** [comment]
I was planning to write more or less this same proposal today for highlighting the method the caret position is in when I found this. 

A couple of my thoughts: 
One key use case for me would be for find operations across files (in refactoring etc) where you quickly move around a lot of script files, and it would be very helpful to see at all times in what methods your found instances are using this. 

As for the highlight of the current method in the list I would go with the current selection highlight, the same that is used when you click a method to go to it (and that then currently remains even if you move away from the method). This as it would then be consistent with the file list above where this selection style is used to show the script file you are currently editing.

Another important feature as well to implement is that when the selection changes in the method list as you move the caret position, the method list should automatically scroll to show this new selection if it is currently not visible. This behavior would also be very helpful in the script file list above to ensure the currently open script file is always visible in the list of files.

I'm glad to see that there's even a PR for implementing it, even if it's now quite old. Is there any chance of perhaps bringing it up to date and getting it merged?

**Comment by Mickeon** [comment]
The major hurdle with the PR is that it [should be language agnostic](https://github.com/godotengine/godot/pull/65338#discussion_r974275175). In this case, that means the code handling this behaviour shouldn't directly be in the text editor itself, and it should be able to change depending on the language, somehow.

**Comment by lostminds** [comment]
Right. Is the method list currently language agnostic? If so it appears to have access to the line position of all the methods in its list. So I’d put the logic there, and just have it listen to caret position changes from the code editor and highlight the first method above the selected line each time the caret line changes.

This will then be incorrect in cases where the caret is past the end of a method but before the next. But I think that’s acceptable. If it’s not, perhaps the method list could figure out the end line positions of methods as well as the starts somehow and get these correct start-end spans for each method in the list?
